src/wallet/coinselection.cpp

```diff
@@ -188,6 +188,156 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
     return result;
 }
 
+/*
+ * The Coin Grinder algorithm is a DFS-based algorithm that deterministically
+ * searches for the minimum-weight input set producing a change output. The
+ * algorithm is closely related to the Branch and Bound algorithm, but will
+ * produce a transaction with a change output instead of aiming to create a
+ * changeless transaction.
+ *
+ * This implementation produces a new input set in every loop iteration. It
+ * uses three operations to transition between input set candidates:
+ *
+ * • ADD: select the UTXO group at the current index into the input set
+ * • SHIFT: deselect the last selected UTXO group, then ADD the next UTXO group
+ * • CUT: deselect the last selected UTXO group, then SHIFT
+ *
+ * E.g. on a UTXO pool consisting of six UTXOs in which the 2nd and 4th are selected, expressing the current selection as a bitmap, the operations would look as follows:
+ *
+ * • ADD:   010100 ⇒ 010110
+ *          the UTXO at the index (here 5th) is additionally selected
+ * • SHIFT: 010100 ⇒ 010010
+ *          the 4th is deselected and the 5th is selected instead
+ * • CUT:   010100 ⇒ 001000
+ *          the third is deselected, the first is deselected and the second is selected
```

Comment:
For consistency, I think this should read:

`the 4th is deselected, the 2nd is deselected and the 3rd is selected`

Reply:
Thanks, I’ll fix that